azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,405 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import os
+
+from azurelinuxagent.common.future import ustr
+
+import azurelinuxagent.common.conf as conf
+import azurelinuxagent.common.logger as logger
+from azurelinuxagent.common.protocol.goal_state import ExtensionsConfig
+from azurelinuxagent.common.utils import fileutil, restutil
+from azurelinuxagent.common.exception import ProtocolError
+from azurelinuxagent.common.utils.shellutil import run_command
+from azurelinuxagent.common.utils.textutil import safe_shlex_split
+
+_INCARNATION_FILE_NAME = "Incarnation"
```

Comment:
Other than VmId, why do we need to use those files as a cache? can't these values be computed during agent initialization.

We had a series of issues about values cached on files because they are basically global state and there are multiple parts of the agent that fetch the goal state at different times. Currently the values to save to files are used only for debugging, but the agent does not read the values from file.

Reply:
The issue is for agent startup. We need to process either the FT or the Fabric GS, but we have no idea which one. Neither document has data indicating which one is more recent, so we have the following logic:
 - If the Fabric incarnation changed, we use Fabric
 - If the FastTrack seqNo changed, we use FastTrack
 - Otherwise, we use the last one processed

The only way to know these upon startup is to save the incarnation, seqNo, and mode. Note that after startup we only save to the files. We don't read them again.

Reply:
But then how is agent update going to work if it cannot compute those values at startup?
Version N doesn't have FT; version N+1 does. How is initialization done on update?

Reply:
If we don't have the cached values on startup, we should be default execute the Fabric GS.

Reply:
BTW, this test hits that scenario - test_startup_first_time

Reply:
Unfortunately, support many times deletes and restarts the agent. If the last GS was FT and then the agent gets removed/re-installed, this is going to break?

Reply:
Also, since the cached values are used only once, can we move the code that reads those files to the init method of the retriever instead of it being part of get_ext_conf?

Reply:
If they delete and remove the agent, then there's nothing we can do. We don't know from the goal states themselves which to use.

Reply:
Removed the cached values